Adds support for optional callbacks onRetriableFailure, debugRequest

joer14 opened the pull request:
Let's try this again...

Adds support for optional callbacks onRetriableFailure, debugRequest
- updates docs
- bumps the minor version
- changes the namespace to `@agilemd/habrok`

@zackliston I'm looking for feedback on everything but in particular:

- method names
- I'm not sure if I'm invoking the callbacks in the correct fashion or if they should be a promise or something?
- the objects I'm passing to the callbacks... do they make sense?
- I didn't write any tests but will attempt to once this stuff looks okay. 
- the optional arguments are passed to the request object. We could instead pass them to the main [configuration object](https://github.com/agilemd/habrok-js#optional-arguments) as optional arguments. 

Sorry about all the questions!

Commit:
Adds support for optional callbacks onRetriableFailure, debugRequest

Also,
- updates docs
- bumps the minor version
- changes the namespace to `@agilemd/habrok`

Commit:
Corrects package version change from patch bump to minor bump

Commit:
Adds proposed changes the changelog (this may be premature)

Commit:
Fixes bug with debugRequest

zackliston commented on lib/Habrok.js:
```diff
@@ -78,13 +78,17 @@ Habrok.prototype.request = function request(raw, opts) {
       if ((!err && self.RETRY_CODES.indexOf(statusCode) === -1) || attempt >= self.RETRIES) {
         debug('fail', req);
-
+        if (has(options, 'debugRequest')) options.debugRequest({ attempt });
```
Use brackets with if statements. Surprised that didn't show up in ESLint

```js
if (has(options, 'debugRequest')) {
  options.debugRequest({ attempt })
}

zackliston replied:
But otherwise looks good. I think the arguments are appropriate. Let's talk about when it is appropriate to use an object for the args and when you should send individual args.

zackliston commented on lib/Habrok.js:
```diff
@@ -78,13 +78,17 @@ Habrok.prototype.request = function request(raw, opts) {
+          if (has(options, 'onRetriableFailure')) {
+            const returnedError = err || boom.create(statusCode, null, body);
+            options.onRetriableFailure(returnedError);
+          }
```
Don't be afraid of whitespace. Optimize for readability. Try putting a new line after the if statement, see how it feels.

zackliston commented on lib/Habrok.js:
```diff
@@ -78,13 +78,17 @@ Habrok.prototype.request = function request(raw, opts) {
         () => {
           debug('retry', raw);
+          if (has(options, 'onRetriableFailure')) {
+            const returnedError = err || boom.create(statusCode, null, body);
```
Is boom.create a method? I don't see documentation, what does it expect? What is the null here for?
https://www.npmjs.com/package/boom

zackliston commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
```
Good work updating the Changelog. We should do this in all of our projects.

zackliston commented on lib/Habrok.js:
```diff
@@ -78,13 +78,17 @@ Habrok.prototype.request = function request(raw, opts) {
           debug('retry', raw);
+          if (has(options, 'onRetriableFailure')) {
+            const returnedError = err || boom.create(statusCode, null, body);
+            options.onRetriableFailure(returnedError);
```
Hmm, what do you think about just calling it `onRetry` the failure isn't really adding much info since we already know it only retries on failures.

The most important piece of information we want here is the status code. We don't necessarily want to pull that out every time, it would be nice to just have it.

`onRetry(statusCode, error)`

What do you think about that signature?

zackliston reviewed (changes requested):
A couple comments. Remember our flow. Docs/test/write.

In this case we are reviewing the docs and implementation at the same time. I understand why you didn't write tests yet, you were not sure that the methods would stay the same based on my review; however, in this case it would have probably been best to just review the docs first, then you could go straight to tests and then implementation and we confirmed the docs.

A note for tests. We don't use them everywhere, so you have to know when you need to implement new tests. A good rule of thumb is to write tests in projects that already implement them. Never decrease test coverage.

zackliston commented:
@joer14 Review complete. Ready for changes.

Commit:
refactor onRetry, debugRequest

- rename 'onRetriableFailure'

Commit:
Updates readme with docs on onRetry, debugRequest

- rename onRetriableFailure to onRetry in changelog

Commit:
Adds test for debugRequest where we reach max retries

Commit:
Adds unit tests for onRetry

- adds test for checking if onRetry is called on every retry
- adds test for checking that onRetry receives the correct status codes as an argument
- adds test for checking that we stop retrying after success, but before max retries.

Commit:
Updates unit tests for success case

- adds debugRequest call when successful
- removes catch statement from tests where we succeed (so any unexpected errors aren't caught)

joer14 commented:
Hey @zackliston 

I think I'm ready for your final review on this. I added new tests and they are passing. Thank you!

zackliston commented on lib/Habrok.js:
```diff
@@ -78,13 +79,17 @@ Habrok.prototype.request = function request(raw, opts) {
     } else if (errorCode === ECONNRESET || statusCode > 299) {
       if ((!err && self.RETRY_CODES.indexOf(statusCode) === -1) || attempt >= self.RETRIES) {
         debug('fail', req);
+        debugRequest({ attempts: attempt });
```
Just one suggest change. I would move all calls of debugRequest to the end as to ensure it is always called.

For example

```js
return resolve({
  statusCode: res.statusCode,
  headers: res.headers,
  body
})
.then((out) => {
  debugRequest({ attempts: attempt });
  return out;
})
.catch((err) => {
  debugRequest({ attempts: attempt });
  throw err;
});
```

joer14 replied:
I think this is a good suggestion, but since the current way is working, and I couldn't figure out how to get this working, and we have an deadline coming up, I'm going to not change to this, if that is alright. 

Ready for your final review / merge if you are. And then we can publish to npm?

zackliston replied:
(thumbsup)

zackliston reviewed (changes requested):
Looks great. Just fix spacing issues in tests file and see comment then let's publish this bad boy.

Commit:
cleans up spacing in tests
